compiler/src/Canvas.js

```diff
@@ -46,6 +46,11 @@ function Canvas(id, w, h, wString, hString) {
 
 // add layer to a canvas
 function addLayer(layer) {
+    console.log(
+        `adding new layer in canvas: ${this.id} with id: ${this.layers.length}`
+    );
+    layer.id = this.id;
+    layer.canvasId = this.layers.length;
```

Comment:
Is `layer.id` and `layer.canvasId` absolutely necessary?

Reply:
Yes these are needed to specify the `dependencies` array, which is an array of layers that are dependent on the current layer, defined by their <canvasId, layerId> unique identifier pair. An example where this is used is in the hierarchical election map example. I should document the `dependencies` attribute as part of the Layer API, which you pointed out in another comment in the PR.